third_party/gtp_ovs/ovs-gtp-patches/2.15/0024-QFI-Support-in-OVS.patch

```diff
@@ -0,0 +1,1021 @@
+@@ -616,9 +640,10 @@ static netdev_tx_t gtp_dev_xmit_fb(struct sk_buff *skb, struct net_device *dev)
+ 		netdev_dbg(dev, "packet with opt len %d", info->options_len);
+ 		if (info->options_len == 0) {
+ 			if (info->key.tun_flags & TUNNEL_OAM) {
+-			   set_qfi = 9;
++			   set_qfi = info->key.qfi;
+ 			}
+-			gtp1_push_header(dev, skb, tunnel_id_to_key32(info->key.tun_id), set_qfi);
++			struct gtpu_metadata *opts = ip_tunnel_info_opts(info);
++			gtp1_push_header(dev, skb, opts, tunnel_id_to_key32(info->key.tun_id), set_qfi);
+ 		} else if (info->key.tun_flags & TUNNEL_GTPU_OPT) {
+ 				struct gtpu_metadata *opts = ip_tunnel_info_opts(info);
+ 				__be32 tid = tunnel_id_to_key32(info->key.tun_id);
+diff --git a/datapath/linux/compat/include/linux/gtp.h b/datapath/linux/compat/include/linux/gtp.h
+index 57d7a128e..24310cc31 100644
+--- a/datapath/linux/compat/include/linux/gtp.h
++++ b/datapath/linux/compat/include/linux/gtp.h
+@@ -8,13 +8,16 @@
+ #endif
+ 
+ enum {
+-	GTP_METADATA_V1
++	GTP_METADATA_V1,
++	GTP_METADATA_EXT_HDR_DATA
+ };
+ 
+ struct gtpu_metadata {
+ 	__u8	ver;
+ 	__u8	flags;
+ 	__u8	type;
++	__u8    header_data_len;
++	__u8    data[];
+ };
+ 
+ enum {
+diff --git a/datapath/linux/compat/include/linux/openvswitch.h b/datapath/linux/compat/include/linux/openvswitch.h
+index ae7772ba3..50d476411 100644
+--- a/datapath/linux/compat/include/linux/openvswitch.h
++++ b/datapath/linux/compat/include/linux/openvswitch.h
+@@ -415,6 +415,7 @@ enum ovs_tunnel_key_attr {
+ 	OVS_TUNNEL_KEY_ATTR_PAD,
+ 	OVS_TUNNEL_KEY_ATTR_ERSPAN_OPTS,	/* struct erspan_metadata */
+ 	OVS_TUNNEL_KEY_ATTR_GTPU_OPTS,		/* struct gtpu_metadata */
++	OVS_TUNNEL_KEY_ATTR_QFI,                /* OVS_TUNNEL_KEY_ATTR_QFI */
+ 	__OVS_TUNNEL_KEY_ATTR_MAX
+ };
+ 
+diff --git a/datapath/linux/compat/include/net/ip_tunnels.h b/datapath/linux/compat/include/net/ip_tunnels.h
+index 4cfaa50c8..95563ef6a 100644
+--- a/datapath/linux/compat/include/net/ip_tunnels.h
++++ b/datapath/linux/compat/include/net/ip_tunnels.h
+@@ -107,6 +107,7 @@ void rpl_ip_tunnel_xmit(struct sk_buff *skb, struct net_device *dev,
+ #define TUNNEL_NOCACHE		__cpu_to_be16(0x2000)
+ #define TUNNEL_ERSPAN_OPT	__cpu_to_be16(0x4000)
+ #define TUNNEL_GTPU_OPT		__cpu_to_be16(0x8000)
++#define TUNNEL_QFI              __cpu_to_be16(0x10000
+ 
+ #undef TUNNEL_OPTIONS_PRESENT
+ #define TUNNEL_OPTIONS_PRESENT \
+@@ -132,7 +133,7 @@ struct tnl_ptk_info {
+ #define IPTUNNEL_ERR_TIMEO	(30*HZ)
+ 
+ /* Used to memset ip_tunnel padding. */
+-#define IP_TUNNEL_KEY_SIZE	offsetofend(struct ip_tunnel_key, tp_dst)
++#define IP_TUNNEL_KEY_SIZE	offsetofend(struct ip_tunnel_key, qfi)
+ 
+ /* Used to memset ipv4 address padding. */
+ #define IP_TUNNEL_KEY_IPV4_PAD	offsetofend(struct ip_tunnel_key, u.ipv4.dst)
+@@ -158,6 +159,7 @@ struct ip_tunnel_key {
+ 	__be32                  label;          /* Flow Label for IPv6 */
+ 	__be16			tp_src;
+ 	__be16			tp_dst;
++	u8                      qfi;
```

Comment:
Lets pass qfi using struct gtpu_metadata{}

Reply:
Changed it.